### DISCUSSION TITLE
Error 500 - connection closed on first call only.

### DISCUSSION THREAD

**Discussion by milsman2** [fastapi/fastapi#6396] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
Session:
from sqlalchemy.ext.asyncio import AsyncSession, create_async_engine
from sqlalchemy.orm import sessionmaker
from app.core.config import settings

engine = create_async_engine(settings.async_database_url)

async_session = sessionmaker(engine, class_=AsyncSession, expire_on_commit=False)

async def get_db():
    async with async_session() as session:
        yield session
---
Endpoint:

@router.get("/all-cast-iron/", status_code=200, response_model=CastIronSearchResults)
async def get_user_items(*, db: AsyncSession = Depends(get_db)) -> Any:
    """
    Fetch all cast iron
    """
    all_cast_iron = await crud.cast_iron.get_multi(db=db)
    print(all_cast_iron)
    if not all_cast_iron:
        return {"results": list()}
    return {"results": list(all_cast_iron)}

CRUD:

from typing import Any, Dict, Generic, List, Optional, Type, TypeVar, Union
from sqlalchemy import select

from fastapi.encoders import jsonable_encoder
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.base_class import Base

ModelType = TypeVar("ModelType", bound=Base)
CreateSchemaType = TypeVar("CreateSchemaType", bound=BaseModel)
UpdateSchemaType = TypeVar("UpdateSchemaType", bound=BaseModel)


class CRUDBase(Generic[ModelType, CreateSchemaType, UpdateSchemaType]):
    def __init__(self, model: Type[ModelType]):
        """
        CRUD object with default methods to Create, Read, Update, Delete (CRUD).
        **Parameters**
        * `model`: A SQLAlchemy model class
        * `schema`: A Pydantic model (schema) class
        """
        self.model = model

    async def get(self, db: AsyncSession, id: int) -> Optional[ModelType]:
        response = await db.scalars(select(self.model).where(self.model.id == id))
        return response.first()

    async def get_multi(self, db: AsyncSession) -> List[ModelType]:
        response = await db.scalars(select(self.model).order_by(self.model.id))
        return response.all()
---
Standard Kubernetes deployment:
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: tpctk-fastapi
  namespace: the-precast-toolkit-fs
spec:
  selector:
    matchLabels:
      app: tpctk-fastapi
  replicas: 1
  template:
    metadata:
      labels:
        app: tpctk-fastapi
    spec:
      containers:
      - name: tpctk-fastapi
        image: milsman2/tpctk-fs-async-backend:v0.3
        ports:
        - containerPort: 8000
          name: tpctk-fastapi
        env:
        - name: DB_USER
          valueFrom:
            secretKeyRef:
              name: fastapi-secret
              key: database_user
        - name: DB_PWD
          valueFrom:
            secretKeyRef:
              name: fastapi-secret
              key: database_password
        - name: DB_HOST
          valueFrom:
            secretKeyRef:
              name: fastapi-secret
              key: database_host
        - name: DB_NAME
          valueFrom:
            secretKeyRef:
              name: fastapi-secret
              key: database_name
      nodeSelector:
        nodeCategory: worker
```

### Description

For some reason, I get a 500 error code on the first request to my fastAPI server based on my custom Docker image running in Kubernetes. I only get this code back the first time and every time after that works just fine. I also see that in my logs from the pod, the error reads out as a connection closed issue. 

### Operating System

Linux, Other

### Operating System Details

This is being run as a custom Docker image on my k3s/bare metal cluster.

### FastAPI Version

0.75.2

### Python Version

Python 3.8.10

### Additional Context

I am running everything using SQLAlchemy ORM and all calls are async.

**Comment by milsman2** [comment]
Full code that this us coming from can be found at https://github.com/milsman2/the-precast-toolkit-fs.

**Comment by milsman2** [comment]
Update: after the program has been up for a while, I still see the following error on first call to the API - 

`sqlalchemy.exc.InterfaceError: (sqlalchemy.dialects.postgresql.asyncpg.InterfaceError) <class 'asyncpg.exceptions._base.InterfaceError'>: connection is closed`

**Comment by milsman2** [comment]
I added `pool_pre_pring=True` to the args for `engine = create_async_engine(settings.async_database_url)` and it seems to be working now. I would like to verify if someone can that this is not due to an underlying issue with the other parts of my code.

**Comment by dmartin-gh** [comment]
@milsman2 I am also encountering this. I will try the `pool_pre_ping = True` option as you suggested, but given I cannot find any other reports of this error I worry I may have the same underlying issue lingering in the service. Did you find out anything else?

**Comment by 9en9i** [comment]
had the same problem. the backend was in a container. there was no problem on the server running docker, after moving to kubernetes there were problems. the second answer helped:

https://stackoverflow.com/questions/70468354/fastapi-sqlalchemy-connection-was-closed-in-the-middle-of-operation/70470805#70470805
